### PR TITLE
[Docs] translate logging design

### PR DESCRIPTION
**Pull request by C1rN09**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: docs/en/design/logging.md [hunk 1]
```diff
@@ -1,3 +1,448 @@
...
+
+- Configuring statistical methods in config files. For example, you can configure to count losses by averaging across the whole training process instead of by smoothing in a sliding window manner
+- Querying training states (iterations, epochs, etc.) anywhere in any module
+- Switching the log saving method between multi-process and master process only
+
+![image](https://user-images.githubusercontent.com/57566630/163441489-47999f3a-3259-44ab-949c-77a8a599faa5.png)
+
+HistoryBuffer encapsules losses, learning rates and some other scalars during training. These HistoryBuffers along with some other runtime info are managed by MessageHub in key-value pairs. The data are then formatted by LogProcessor and exported to various visualization backends by [LoggerHook](mmengine.hook.LoggerHook). **In most cases, you don't need to care about data flow in this process. You can simply configure the LogProcessor to choose from statistical methods.**  Before diving into the design of logging system in MMEngine, you may want to read through [logging tutorial](../tutorials/logging.md) to familiarize yourself with some basic use cases.
```
Comment (zhouzaida) [inline docs/en/design/logging.md:13]:
```suggestion
HistoryBuffer encapsulates losses, learning rates, and some other scalars during training. These HistoryBuffers along with some other runtime info are managed by MessageHub in key-value pairs. The data are then formatted by LogProcessor and exported to various visualization backends by [LoggerHook](mmengine.hook.LoggerHook). **In most cases, you don't need to care about data flow in this process. You can simply configure the LogProcessor to choose from statistical methods.**  Before diving into the design of logging system in MMEngine, you may want to read through [logging tutorial](../advanced_tutorials/logging.md) to familiarize yourself with some basic use cases.
```

File: docs/en/design/logging.md [hunk 2]
```diff
@@ -1,3 +1,448 @@
...
+message_hub = MessageHub.get_instance('task')
+message_hub.update_info('iter', 1)
+message_hub.get_info('iter')  # 1
+message_hub.update_info('iter', 2)
+message_hub.get_info('iter')  # 2, overwritten by the above command
+```
+
+### Share MessageHub across modules
+
+During the execution of a runner, different modules receive and post data through MessageHub. Then, [RuntimeInfoHook](mmengine.hooks.RuntimeInfoHook) gathers data such as losses and learning rates before exporting them to user defined backends(Tensorboard, WandB, etc). Following is an example to show the communication between logger hook and other modules.
```
Comment (zhouzaida) [inline docs/en/design/logging.md:211]:
```suggestion
During the execution of a runner, different modules receive and post data through MessageHub. Then, [RuntimeInfoHook](mmengine.hooks.RuntimeInfoHook) gathers data such as losses and learning rates before exporting them to user defined backends (Tensorboard, WandB, etc). Following is an example to show the communication between logger hook and other modules.
```

File: docs/en/design/logging.md [hunk 3]
```diff
@@ -1,3 +1,448 @@
...
+log_processor = dict(
+    window_size=10,
+    by_epoch=True,
+    custom_cfg=[
+        dict(data_src='loss',
+             method_name='mean',
+             window_size=100)])
+```
+
+In this configuration, LogProcessor will overwrite the default window size 10 by a larger window size 100 and output the mean value to 'loss' field in logs.
```
Comment (zhouzaida) [inline docs/en/design/logging.md:338]:
```suggestion
In this configuration, LogProcessor will overwrite the default window size 10 by a larger window size 100 and output the mean value to 'loss' field in logs.
```

File: docs/en/design/logging.md [hunk 4]
```diff
@@ -1,3 +1,448 @@
 # Logging
 
-Coming soon. Please refer to [chinese documentation](https://mmengine.readthedocs.io/zh_CN/latest/design/logging.html).
+## Overview
+
+[Runner](./runner.md) produces a great amount of logs during execution. These logs include but not limited to dataset information, model initialization, learning rates and losses. In order to make these logs easily accessed by users, we designed [MessageHub](mmengine.logging.MessageHub), [HistoryBuffer](mmengine.logging.HistoryBuffer), [LogProcessor](mmengine.runner.LogProcessor) and [MMLogger](mmengine.logging.MMLogger) in MMEngine, which enables:
```
Comment (Harold-lkk) [inline docs/en/design/logging.md:5]:
```suggestion
[Runner](./runner.md) produces amounts of logs during execution. These logs include dataset information, model initialization, learning rates, losses, etc. In order to make these logs easily accessed by users, MMEngine designs [MessageHub](mmengine.logging.MessageHub), [HistoryBuffer](mmengine.logging.HistoryBuffer), [LogProcessor](mmengine.runner.LogProcessor) and [MMLogger](mmengine.logging.MMLogger), which enable:
```

File: docs/en/design/logging.md [hunk 5]
```diff
@@ -1,3 +1,448 @@
...
+
+- Configuring statistical methods in config files. For example, you can configure to count losses by averaging across the whole training process instead of by smoothing in a sliding window manner
+- Querying training states (iterations, epochs, etc.) anywhere in any module
+- Switching the log saving method between multi-process and master process only
```
Comment (Harold-lkk) [inline docs/en/design/logging.md:9]:
```suggestion
- Configure statistical methods in config files. For example, losses can be globally averaged or smoothed by a sliding window.
- Query training states (iterations, epochs, etc.) in any module
- Configure whether save the multi-process log or not during distributed training.
```

File: docs/en/design/logging.md [hunk 6]
```diff
@@ -1,3 +1,448 @@
...
+
+![image](https://user-images.githubusercontent.com/57566630/163441489-47999f3a-3259-44ab-949c-77a8a599faa5.png)
+
+HistoryBuffer encapsulates losses, learning rates, and some other scalars during training. These HistoryBuffers along with some other runtime info are managed by MessageHub in key-value pairs. The data are then formatted by LogProcessor and exported to various visualization backends by [LoggerHook](mmengine.hook.LoggerHook). **In most cases, you don't need to care about data flow in this process. You can simply configure the LogProcessor to choose from statistical methods.**  Before diving into the design of logging system in MMEngine, you may want to read through [logging tutorial](../advanced_tutorials/logging.md) to familiarize yourself with some basic use cases.
```
Comment (Harold-lkk) [inline docs/en/design/logging.md:13]:
```suggestion
Each scalar (losses, learning rates, etc.) during training is encapsulated by HistoryBuffer, managed by MessageHub in key-value pairs, formatted by LogProcessor and then exported to various visualization backends by [LoggerHook](mmengine.hook.LoggerHook). **In most cases, statistical methods of these scalars can be configured through the LogProcessor without understanding the data flow.**  Before diving into the design of the logging system, please read through [logging tutorial](../advanced_tutorials/logging.md) first for familiarizing basic use cases.
```

File: docs/en/design/logging.md [hunk 7]
```diff
@@ -1,3 +1,448 @@
...
+
+## HistoryBuffer
+
+MMEngine provides HistoryBuffer class to record the trajectory of training. Scalars such as losses, learning rates and iteration time are saved to corresponding HistoryBuffer as an element in an array. In most cases, HistoryBuffer is used as an internal class. It works with [MessageHub](mmengine.logging.MessageHub), LoggerHook and [LogProcessor](mmengine.runner.LogProcessor) to make training log configurable.
+
+However, HistoryBuffer can also be used alone, which enables users to manage their training logs and do various statistics in an easy manner. Therefore, we will first introduce the usage of HistoryBuffer in the following section. The association between HistoryBuffer and MessageHub will be introduced later in MessageHub section.
```
Comment (Harold-lkk) [inline docs/en/design/logging.md:19]:
```suggestion
`HistoryBuffer` records the history of the corresponding scalar such as losses, learning rates, and iteration time in an array and as an internal class, works with [MessageHub](mmengine.logging.MessageHub), LoggerHook and [LogProcessor](mmengine.runner.LogProcessor) to make training log configurable. Meanwhile, HistoryBuffer can also be used alone, which enables users to manage their training logs and do various statistics in an easy manner.
We will first introduce the usage of HistoryBuffer in the following section. The association between HistoryBuffer and MessageHub will be introduced later in the MessageHub section.
```

File: docs/en/design/logging.md [hunk 8]
```diff
@@ -1,3 +1,448 @@
...
+
+### Initialize HistoryBuffer
+
+HistoryBuffer accepts 2 arguments, named `log_history` and `count_history`, for initialization. `log_history` records the trajectory of the training. For example, if the losses in the previous 3 iterations are 0.3, 0.2, 0.1 respectively, then we have `log_history=[0.3, 0.2, 0.1]`. `count_history` is more indirect. It controls the statistical granularity and is only applied when counting the average. Take the above example, if we count the average loss across iterations, then we have `count_history=[1, 1, 1]`. Instead, if we count the average loss across images with `batch_size=8`, then we have `count_history=[8, 8, 8]`.
```
Comment (Harold-lkk) [inline docs/en/design/logging.md:23]:
```suggestion
### HistoryBuffer Initialization

HistoryBuffer accepts `log_history`, `count_history` and `max_length` for initialization. 
* `log_history` records the history of the scaler. For example, if the loss in the previous 3 iterations is 0.3, 0.2, 0.1 respectively, the `log_history` will be [0.3, 0.2, 0.1]. 
* `count_history` controls the statistical granularity and will be used when counting the average. Take the above example, if we count the average loss across iterations, we have `count_history=[1, 1, 1]`. Instead, if we count the average loss across images with `batch_size=8`, then we have `count_history=[8, 8, 8]`.
* `max_length`controls the maximum length of the history. If the length of `log_history` and `count_history` exceeds `max_length`, the earliest elements will be removed.
Besides, We can access the trajectory of the data through `history_buffer.data`.
```

File: docs/en/design/logging.md [hunk 9]
```diff
@@ -1,3 +1,448 @@
...
+history_buffer = HistoryBuffer([1, 2, 3], [1, 2, 3])  # Init with lists
+log_history, count_history = history_buffer.data
+# [1 2 3] [1 2 3]
+history_buffer = HistoryBuffer([1, 2, 3], [1, 2, 3], max_length=2)
+# The length of history buffer(3) exceeds the max_length(2), the first few elements will be ignored.
+log_history, count_history = history_buffer.data
+# [2 3] [2 3]
+```
+
+We can access the trajectory of the data through `history_buffer.data`. Besides, we can also set the maximum length of the HistoryBuffer. When the length of a HistoryBuffer exceeds its maximum length, the earliest few data items will be dropped.
```
Comment (Harold-lkk) [inline docs/en/design/logging.md:40]:
```suggestion
```

File: docs/en/design/logging.md [hunk 10]
```diff
@@ -1,3 +1,448 @@
...
+
+### Update HistoryBuffer
+
+We can update the HistoryBuffer with `update` method. It accepts 2 arguments, where the first updates `log_history` while the second updates `count_history`.
```
Comment (Harold-lkk) [inline docs/en/design/logging.md:44]:
```suggestion
### HistoryBuffer Update

We can update the `log_history` and `count_history` through `HistoryBuffer.`update(log_history, count_history)`.
```

File: docs/en/design/logging.md [hunk 11]
```diff
@@ -1,3 +1,448 @@
...
+history_buffer = HistoryBuffer([1, 2, 3], [1, 1, 1])
+history_buffer.update(4)  # count default to 1
+log_history, count_history = history_buffer.data
+# [1, 2, 3, 4] [1, 1, 1, 1]
+history_buffer.update(5, 2)
+log_history, count_history = history_buffer.data
+# [1, 2, 3, 4, 5] [1, 1, 1, 1, 2]
+```
+
+### Basic statistical methods
```
Comment (Harold-lkk) [inline docs/en/design/logging.md:56]:
```suggestion
### Basic Statistical Methods
```

File: docs/en/design/logging.md [hunk 12]
```diff
@@ -1,3 +1,448 @@
...
+history_buffer = HistoryBuffer([1, 2, 3], [1, 1, 1])
+history_buffer.statistics('mean')
+# 2, as global mean
+history_buffer.statistics('mean', 2)
+# 2.5, as the mean of [2, 3]
+history_buffer.statistics('mean', 2, 3)
+# Error! mismatch arguments given to `mean(window_size)`
+history_buffer.statistics('data')
+# Error! `data` method not registered
+```
```
Comment (Harold-lkk) [inline docs/en/design/logging.md:125]:
```suggestion
### Statistical Methods Invoking

Statistical methods can be accessed through `HistoryBuffer.statistics` with method name and arguments. The `name` parameter should be a registered method name (i.e. built-in methods like `min` and `max`), while arguments should be the corresponding method's arguments.

\```python
history_buffer = HistoryBuffer([1, 2, 3], [1, 1, 1])
history_buffer.statistics('mean')
# 2, as global mean
history_buffer.statistics('mean', 2)
# 2.5, as the mean of [2, 3]
history_buffer.statistics('mean', 2, 3)
# Error! mismatch arguments given to `mean(window_size)`
history_buffer.statistics('data')
# Error! `data` method not registered
\```

### Statistical Methods Registration

Custom statistical methods can be registered through `@HistoryBuffer.register_statistics`.

\```python
from mmengine.logging import HistoryBuffer
import numpy as np


@HistoryBuffer.register_statistics
def weighted_mean(self, window_size, weight):
    assert len(weight) == window_size
    return (self._log_history[-window_size:] * np.array(weight)).sum() / \
            self._count_history[-window_size:]


history_buffer = HistoryBuffer([1, 2], [1, 1])
history_buffer.statistics('weighted_mean', 2, [2, 1])  # get (2 * 1 + 1 * 2) / (1 + 1)
\```

```

File: docs/en/design/logging.md [hunk 13]
```diff
@@ -1,3 +1,448 @@
...
+
+### Use cases
```
Comment (Harold-lkk) [inline docs/en/design/logging.md:127]:
```suggestion
### Use Cases
```

File: docs/en/design/logging.md (hunk 12 above)
Comment (C1rN09) [reply docs/en/design/logging.md:125]: Mark, Chinese docs should also be changed

**Review by zhouzaida — APPROVED** [review]

**Pull request merged**